docs/source/examples/checkpointer.rst

```diff
@@ -0,0 +1,451 @@
+        # if we're restarting a previous run, we need to specify
+        # the file with the checkpoint state. More on this in the
+        # next section
+        recipe_checkpoint: null
+
+        # dir for saving the output checkpoints. Usually set
+        # to be the same as checkpoint_dir
+        output_dir: <checkpoint_dir>
+
+        # model_type which specifies how to convert the state_dict
+        # into a format which TorchTune understands
+        model_type: LLAMA2
+
+    # set to True if restarting training
+    resume_from_checkpoint: False
+
+|
+
+**TorchTuneCheckpointer**
+
+This checkpointer reads and writes checkpoints in a format that is compatible with TorchTune's
+model definition. This does not perform any state_dict conversions and is currently used either
+for testing or for loading quantized models for generation.
+
+|
+
+
+Intermediate vs Final Checkpoints
+---------------------------------
+
+TorchTune Checkpointers support two checkpointing scenarios:
+
+**End-of-training Checkpointing**
+
+The model weights at the end of a completed training
+run are written out to file. The checkpointer ensures that the output checkpoint
+files have the same keys as the input checkpoint file used to begin training. The
+checkpointer also ensures that the keys are paritioned across the same number of
+files as the original checkpoint. The output state dict has the following
+standard format:
+
+  .. code-block:: python
+
+            {
+                "key_1": weight_1,
+                "key_2": weight_2,
+                ...
+            }
+
+
+**Mid-training Chekpointing**.
+
+If checkpointing in the middle of training, the output checkpoint needs to store additional
+information to ensure that subsequent training runs can be correctly restarted. In addition to
+the model checkpoint files, we output a ``recipe_state.pt`` file for intermediate
+checkpoints. These are currently output at the end of each epoch, and contain information
+such as optimizer state, number of epochs completed etc.
+
+To prevent us from flooding ``output_dir`` with checkpoint files, the recipe state is
+overwritten at the end of each epoch.
+
+The output state dicts have the following formats:
+
+    .. code-block:: python
+
+        Model:
+            {
+                "key_1": weight_1,
+                "key_2": weight_2,
+                ...
+            }
+
+        Recipe State:
+            {
+                "optimizer": ...,
+                "epoch": ...,
+                ...
+            }
+
+To restart from a previous checkpoint file, you'll need to make the following changes
+to the config file
+
+.. code-block:: yaml
+
+    checkpointer:
+
+        # checkpointer to use
+        _component_: torchtune.utils.FullModelHFCheckpointer
+
+        # directory with the checkpoint files
+        # this should match the output_dir above
+        checkpoint_dir: <checkpoint_dir>
+
+        # checkpoint files. These refer to intermediate checkpoints
```

Comment:
Noted that these are the intermediate checkpoint files generated, though might be clearer to explicitly mention what users need to change when loading in from a mid-training checkpoint. i.e. explicitly mention "replace the original checkpoint files with the ones saved during mid training". Might be obvious but feel like changing config stuff is pretty easy to forget.